lightning/src/routing/router.rs

```diff
@@ -333,79 +549,229 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+						// hop, if the last hop was provided via a BOLT 11 invoice (though we
+						// should be able to extend it further as BOLT 11 does have feature
+						// flags for the last hop node itself).
+						assert!(ordered_hops.last().unwrap().route_hop.pubkey == *target);
+					}
+
+					if new_entry.available_liquidity_msat > new_entry.following_hops_fees_msat {
+						// How much value a path can transfer is how much the weakest link can transfer after paying
+						// for the use of the following channels.
+						path_bottleneck_msat = cmp::min(path_bottleneck_msat, new_entry.available_liquidity_msat - new_entry.following_hops_fees_msat);
 					} else {
-						res.last_mut().unwrap().node_features = NodeFeatures::empty();
+						// TODO: explain
+						path_bottleneck_msat = cmp::min(path_bottleneck_msat, new_entry.available_liquidity_msat / 10);
+					}
+
+					if ordered_hops.last().unwrap().route_hop.pubkey == *target {
+						break;
 					}
-				} else {
-					// We should be able to fill in features for everything except the last
-					// hop, if the last hop was provided via a BOLT 11 invoice (though we
-					// should be able to extend it further as BOLT 11 does have feature
-					// flags for the last hop node itself).
-					assert!(res.last().unwrap().pubkey == *target);
+
+					new_entry = match dist.remove(&ordered_hops.last().unwrap().route_hop.pubkey) {
+						Some(payment_hop) => payment_hop,
+						None => {
+							// If we can't reach a given node, something wen't wrong during path traverse.
+							// Stop looking for more paths here because next iterations will face the exact same issue.
+							break 'paths_collection;
+						}
+					};
+					// We "propagate" the fees one hop backward (topologically) here, so that fees paid on the
+					// current channels are associated with the previous channel (where they will be paid).
+					ordered_hops.last_mut().unwrap().route_hop.fee_msat = new_entry.hop_use_fee_msat;
+					ordered_hops.last_mut().unwrap().route_hop.cltv_expiry_delta = new_entry.route_hop.cltv_expiry_delta;
+					ordered_hops.push(new_entry.clone());
 				}
-				if res.last().unwrap().pubkey == *target {
-					break;
+				ordered_hops.last_mut().unwrap().route_hop.fee_msat = final_value_msat;
+				ordered_hops.last_mut().unwrap().route_hop.cltv_expiry_delta = final_cltv;
+
+				let mut payment_path = PaymentPath {hops: ordered_hops};
+				// Since a path allows to transfer as much value as the smallest channel it has ("bottleneck"),
+				// we should recompute the fees so sender HTLC don't overpay fees when traversing larger channels than the bottleneck.
+				// This may happen because when we were selecting those channels we were not aware how much value this path
+				// will transfer, and the relative fee for them might have been computed considering a larger value.
+				payment_path.update_value_and_recompute_fees(path_bottleneck_msat);
```

Comment:
If we agree as value-as-final-value-msat shouldn't this be called `update_amount_and_recompute_fees` ?

If so, I think you should switch "a larger value" to a "larger amount". Value is static, amount is function of channel policy, bottleneck, bookkeeping, ...

Reply:
Why do you think the value is static? We're talking about per-path value here.
If we have 10 paths per route, I might change the value each of the paths transfers (reallocate)

Reply:
Right, I interpreted "value" as defined here : https://github.com/rust-bitcoin/rust-lightning/pull/646#discussion_r504718057, but then I realized we don't a naming for per-path value. Maybe per-path amount ?